include/open62541pp/detail/Result.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include <cassert>
+
+#include "open62541pp/ErrorHandling.h"
+#include "open62541pp/types/Builtin.h"  // StatusCode
+
+namespace opcua::detail {
+
+/**
+ * Represents a bad result stored in `Result`.
+ */
+class BadResult {
+public:
+    constexpr explicit BadResult(StatusCode code) noexcept
+        : code_(code) {
+        assert(code.isBad());
+    }
+
+    constexpr StatusCode code() const noexcept {
+        return code_;
+    }
+
+private:
+    StatusCode code_;
+};
+
+/**
+ * Result encapsulates a result value and a status code.
+ * A result may contain just an error status code, just the return value or both.
+ */
+template <typename T>
+class Result {
```

Comment:
I was thinking about that as well. Do you think the interface of the `Result` class is suitable to make it public?
Using the `useFuture` token should still result in `std::future<T>` and not `std::future<Result<T>>`, do you agree?

Reply:
Hm, yes and no 🤔 

Now I remembered an odd characteristic about OPC UA: According to the [Modeling Best Practices](https://opcfoundation.org/wp-content/uploads/2020/09/OPC-11030-Whitepaper-UA-Modeling-Best-Practices-1.00.00.pdf), application-specific errors are to be transmitted in the following manner:

- Statuscode Uncertain
- The output is only partly valid (at least the first output value is valid, returning an application-specific error code)

In that regard the Result class is a perfect match, as it contains the status code as well as the value. Though I assume this means the following for method calls:

- The Result class might need an additional isUncertain() method for clarity
- The uncertain status code possibility also has to be available in the `useFuture` variant

So in order to support this possibility, I tend towards `std::future<Result<T>>`. What do you think?

Reply:
I didn't know about the best practices paper. Thanks for the link 👍
If we use the `Result` type for service responses, we should use it for the sync functions as well. We could do a non-breaking migration by allowing implicit conversions from `Result<T>` to `T&` but deprecating it. The users code should still work but deprecations will pop up during compiling.

What do you think about `void` results, e.g. from `writeValue`?
1. `void` return type, throw exception on error
2. `Result<void>` return type
3. `StatusCode` return type

Reply:
Good idea with the implicit conversion for compatibility 👍 

Regarding other situations like `writeValue` it looks like also there there might be situations where 2. and 3. make more sense, although the specification is a little fuzzy. The read & write APIs use OPC UA's attribute service set, right? Then this is what I found in the spec:

About Reads:
- [This base page](https://reference.opcfoundation.org/Core/Part4/v104/docs/5.10.2) mentions no uncertain codes, ...
- but the linked [table 178](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=_Ref127606591) refers to common status codes which do have uncertain codes ...
- and there is a chapter about the returned [DataValue](https://reference.opcfoundation.org/Core/Part4/v104/docs/7.7) which also mentions uncertain status codes

About Writes:
- Similar to reads, [the base page](https://reference.opcfoundation.org/Core/Part4/v104/docs/5.10.4) refers to table 178

I don't understand whether these status codes from the base pages are supposed to represent the only valid subset of table 178. However, probably not, because none of them mention any uncertain codes (even not the [Call service](https://reference.opcfoundation.org/Core/Part4/v104/docs/5.11.2)).

For the sake of supporting uncertain codes in all scenarios and having a consistent Result-oriented API, what do you think about 2. `Result<void>`?
And I think implicit conversions from `Result<void>` to `StatusCode` would also be fine in order to support the 3. proposal based on personal preference?

Reply:
We didn't come to a conclusion if we should expose the `Result` class and use it for async callbacks and returned futures yet.

I think that the error handling could be simplified and improved with the `Result` class. Also the sync functions would benefit from this approach. But I don't see the `Result` class to be production-ready now. We might have to iterate on the interface to make it work smoothly. Therefore I propose to stay with the current design, use, test and improve the `Result` class internally and eventually use it in the future as part of the public API.

We can easily adapt to the `Result` class for asynchronous operations and allow different callback signatures:

```cpp
// now
void(StatusCode code);  // void
void(StatusCode code, T& result);  // non-void

// maybe later
void(Result<void> result);  // void
void(Result<T>& result);  // non-void
```

Do you agree @chandryan?

Reply:
Yes that sounds good, also seeing how big this pull request has become 😀 
Maybe I can take a look at making it more mature in a separate PR. Is there anything specific you have in mind, which is missing in the current implementation?

Reply:
Totally, this is getting a monster 😀
I am currently doing a greater refactoring, because we need to move all the contexts of server, client, monitored items, subscriptions, methods to a public header (detail). Only then, we can implement the templated async functions for the subscriptions and monitored items.
Let me finish those refactorings and then discuss how to continue. From there on, we should iterate in smaller steps / PRs.
Do you have something specific in mind, what you would like to work on?

Reply:
I was thinking about focussing on the Result class, as you said "We might have to iterate on the interface to make it work smoothly". From the first look I couldn't spot anything that comes to my mind which I don't like about the interface, though I could give it some more time to think about it.